src/Pos/Context/Functions.hs

```diff
@@ -0,0 +1,115 @@
+import           Universum
+
+import           Pos.Context.Class      (WithNodeContext (..))
+import           Pos.Context.Context    (NodeContext (..), ncGenesisLeaders,
+                                         ncGenesisUtxo, ncStartTime)
+import           Pos.Lrc.Error          (LrcError (..))
+import           Pos.Txp.Toil.Types     (Utxo)
+import           Pos.Types              (EpochIndex, HeaderHash, SlotLeaders)
+import           Pos.Util               (maybeThrow, readTVarConditional)
+
+----------------------------------------------------------------------------
+-- Genesis
+----------------------------------------------------------------------------
+
+genesisUtxoM :: (Functor m, WithNodeContext ssc m) => m Utxo
+genesisUtxoM = ncGenesisUtxo <$> getNodeContext
+
+genesisLeadersM :: (Functor m, WithNodeContext ssc m) => m SlotLeaders
+genesisLeadersM = ncGenesisLeaders <$> getNodeContext
+
+----------------------------------------------------------------------------
+-- Semaphore-related logic
+----------------------------------------------------------------------------
+
+takeBlkSemaphore
+    :: (MonadIO m, WithNodeContext ssc m)
+    => m HeaderHash
+takeBlkSemaphore = liftIO . takeMVar . ncBlkSemaphore =<< getNodeContext
+
+putBlkSemaphore
+    :: (MonadIO m, WithNodeContext ssc m)
+    => HeaderHash -> m ()
+putBlkSemaphore tip = liftIO . flip putMVar tip . ncBlkSemaphore =<< getNodeContext
+
+readBlkSemaphore
+    :: (MonadIO m, WithNodeContext ssc m)
+    => m HeaderHash
+readBlkSemaphore = liftIO . readMVar . ncBlkSemaphore =<< getNodeContext
+
+----------------------------------------------------------------------------
+-- LRC synchronization
+----------------------------------------------------------------------------
+
+-- | Block until LRC data is available for given epoch.
+waitLrc
+    :: (MonadIO m, WithNodeContext ssc m)
+    => EpochIndex -> m ()
+waitLrc epoch = do
+    sync <- ncLrcSync <$> getNodeContext
+    () <$ readTVarConditional ((>= epoch) . snd) sync
+
+lrcActionOnEpoch
+    :: (MonadIO m, WithNodeContext ssc m, MonadThrow m)
+    => EpochIndex
+    -> (EpochIndex -> m (Maybe a))
+    -> m a
+lrcActionOnEpoch epoch =
+    lrcActionOnEpochReason
+        epoch
+        "action on lrcCallOnEpoch couldn't be performed properly"
+
+lrcActionOnEpochReason
+    :: (MonadIO m, WithNodeContext ssc m, MonadThrow m)
+    => EpochIndex
+    -> Text
+    -> (EpochIndex -> m (Maybe a))
+    -> m a
+lrcActionOnEpochReason epoch reason actionDependsOnLrc = do
+    waitLrc epoch
+    actionDependsOnLrc epoch >>= maybeThrow (LrcDataUnknown epoch reason)
+
+----------------------------------------------------------------------------
+-- Misc
+----------------------------------------------------------------------------
+
+-- | Returns node uptime based on current time and 'ncStartTime'.
+getUptime :: (MonadIO m, WithNodeContext ssc m) => m Microsecond
+getUptime = do
+    curTime <- liftIO getCurrentTime
```

Comment:
❗️ I think `UTCTime` is inappropriate for uptime. See: http://chrisdone.com/posts/measuring-duration-in-haskell